### ISSUE TITLE
Publishing a blazor wasm app throws a Mono Linker exception in .NET 7.0 Preview 6

### ISSUE THREAD

**Issue by Stamo-Gochev** [dotnet/linker#2903]

**Comment by vitek-karas** [comment]
I think that we could use the dictionary as recursion detection as well - basically immediately at the beginning add an entry which say `true` for the value and then update it at the end to the correct value.

This would mean that if we hit the recursion case, the method will be treated as "Requires method body scanner" - which it probably does (or at least indirectly). This will fix the SO.

There's a UX problem here though, the code in question is very simple
```C#
		class LambdaWhichMarksItself
		{
			public static void Test ()
			{
				var a = () => {
					RequiresAllOnT<LambdaWhichMarksItself> ();
				};
                        }
                }
```

Accessing this lambda by reflection is obviously not problematic in any way - it has no closure access and it has no interesting parameters and so it acts as effectively a normal method.

Basically the current heuristic fails in this case and produces noise. We'll have to think about potential fixes around this. But that should not block fixing the SO - which is obviously much worse.

**Comment by Stamo-Gochev** [comment]
Hi all, I see that the PR https://github.com/dotnet/linker/pull/2904 is now merged, but there isn't a milestone assigned to the issue - will it be available in .NET 7.0 Preview 7?

**Comment by vitek-karas** [comment]
#2904 is just a test for this - so that it's easier to repro. It is NOT a fix.

The current plan is to fix this for .NET 7, but it won't make it into Preview 7 (too late for that). So hopefully RC1.
